File: cli/src/git.rs
use std::io;
use std::process::Command;

use crate::{Error, Result};

pub fn diff(base: Option<&str>, staged: bool, paths: &[String]) -> Result<Vec<u8>> {
    let mut command = Command::new("git");
    command.arg("diff");

    if let Some(base) = base {
        command.arg(base);
    }
    if staged {
        command.arg("--staged");
    }
    command.arg("--").args(paths);

    let output = command.output().map_err(|error| -> Error {
        if error.kind() == io::ErrorKind::NotFound {
            "git is required".into()
        } else {
            format!("failed to run git diff: {error}").into()
        }
    })?;

    if !output.status.success() {
        let stderr = String::from_utf8_lossy(&output.stderr);
        let message = stderr.trim();

        return Err(match message.is_empty() {
            true => format!("git diff exited with {}", output.status).into(),
            false => message.into(),
        });
    }

    if output.stdout.iter().all(u8::is_ascii_whitespace) {
        return Err("No changes found".into());
    }

    Ok(output.stdout)
}

File: cli/git.go
package main

import (
	"bytes"
	"errors"
	"fmt"
	"os/exec"
	"strings"
)

func diff(base string, staged bool, paths []string) ([]byte, error) {
	out, err := exec.Command("git", diffArgs(base, staged, paths)...).Output()
	if err != nil {
		return nil, diffErr(err)
	}
	if len(bytes.TrimSpace(out)) == 0 {
		return nil, errors.New("No changes found")
	}
	return out, nil
}

func diffArgs(base string, staged bool, paths []string) []string {
	args := []string{"diff"}
	if base != "" {
		args = append(args, base)
	}
	if staged {
		args = append(args, "--staged")
	}
	return append(append(args, "--"), paths...)
}

func diffErr(err error) error {
	if errors.Is(err, exec.ErrNotFound) {
		return errors.New("Git is required")
	}
	var exitErr *exec.ExitError
	if errors.As(err, &exitErr) {
		msg := strings.TrimSpace(string(exitErr.Stderr))
		if msg == "" {
			return fmt.Errorf("Git diff exited with %v", exitErr.ProcessState)
		}
		return errors.New(msg)
	}
	return fmt.Errorf("Run git diff: %w", err)
}
